# Baseline Model
Now that the data is cleaned, it's time to create a baseline model. 
This will be a simple but effective model that provides a benchmark score.
The simple neural model will need to outperform this baseline to be considered a success.

## Setup and Load of the data

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# Load the FULL CLEANED data
file_path = '../data/reviews_cleaned_FULL.csv'
df_clean = pd.read_csv(file_path)

# Drop any rows that might be empty after cleaning
df_clean.dropna(inplace=True)
print(df_clean.head())

                                              review sentiment  \
0  One of the other reviewers has mentioned that ...  positive   
1  A wonderful little production. <br /><br />The...  positive   
2  I thought this was a wonderful way to spend ti...  positive   
3  Basically there's a family where a little boy ...  negative   
4  Petter Mattei's "Love in the Time of Money" is...  positive   

                                      cleaned_review  
0  one of the other reviewer have mention that af...  
1  a wonderful little production the film techniq...  
2  I think this be a wonderful way to spend time ...  
3  basically there be a family where a little boy...  
4  petter mattei 's love in the time of money be ...  


## Split of the Dataset
As a best practice, I will split the data into training set and testing sets.
The model will be trained on 80% of the data and then evaluated on the remaining 20% of unseen data to get a realistic measure of it performance.


In [2]:
# Define  features (X) and  target (y)
x = df_clean['cleaned_review']
y = df_clean['sentiment']

# Split the data
x_train, x_test, y_train , y_test = train_test_split(x, y, test_size=0.2, random_state=42, stratify=y)

print(f"Training set size: {len(x_train)}")
print(f"Testing set size: {len(x_test)}")


Training set size: 40000
Testing set size: 10000


### Vectorize the text with TF-IDF and N-grams
Will fit the vectorizer only on the training data to learn the vocabulary, and then use it to transform both the training and testing data


In [3]:
# Initialize the TF-IDF Vectorizer with the N-gram strategy
tfidf = TfidfVectorizer(ngram_range=(1, 3)) # Use unigrams, bigrams and trigrams

# Fit and transform the training data
x_train_tfidf= tfidf.fit_transform(x_train)

# Only transform the test data (using the vocabulary learned from the training data)
x_test_tfidf = tfidf.transform(x_test)

print("Text data vectorizer succesfully.")

Text data vectorizer succesfully.


In [4]:
# Initialize a simple and powerful logistic regression model
model = LogisticRegression()

# Train the model on our vectorized training data
print("Training the baseline model...")
model.fit(x_train_tfidf, y_train)
print("Model training complete!")


Training the baseline model...
Model training complete!


## Evaluation of the baseline model

In [5]:
# Make predictions on the unseen test data
y_pred = model.predict(x_test_tfidf)

# The final Report 
print("\n Baseline Model Performance")

print(classification_report(y_test, y_pred))


 Baseline Model Performance
              precision    recall  f1-score   support

    negative       0.91      0.88      0.89      5000
    positive       0.89      0.91      0.90      5000

    accuracy                           0.89     10000
   macro avg       0.90      0.89      0.89     10000
weighted avg       0.90      0.89      0.89     10000



## Interpretation of results and conclusion (Full Dataset)

When trained on a full 50.000 review data set (40,000 training/10,000 testing), the baseline model's performance increases dramatically, achieving an outstanding **weighted average F1-score of 0.89**. This a significant 18-point improvement over the 0.71 F1-score achieved on the 1,000-review sample.

### Perfomance Analysis
Unlike the sample run where the model showed a clear bias, the fully trained model is well-balanced and robust. The performance is strong accross both classes:

* **Negative Class:** The model perfoms exceptionally well, with a Precision of 0.91 and a Recall of 0.88, resulting in an F1-Score of 0.89.
* **Positive Class:** Similarly, it achieves a Precision of 0.89 and a Recall of 0.91 for positive reviews, with a final F1-Score of 0.90

This indicates that the model is highly correctly identifying both positive and negative reviews with no significant bias.




## Conclusion

This result is a powerful demonstration of a key machine learning principle: data quantity can dramatically improve model performance. The TF-IDF + N-GRAM approach, when trained on a sufficiently large dataset, has proven to be a high performance, state of art solution for the task.

The baseline score to beat with the simple neural model is now set at a very competitive **F1-Score of 0.89** 